File: Arrays/Q3. Count Pairs Whose Sum is Less than Target/Q3_Count_Pairs_Whose_Sum_is_Less_than_Target.cpp
# include <iostream>
# include <vector>
using namespace std;

int countPairs(int nums[], int n, int target) {
    int cnt = 0; 
    for (int i = 0; i < n - 1; i ++){
        for (int j = i + 1; j < n; j ++){
            if (nums[i] + nums[j] < target){
                cnt ++;
            }
        }
    }
    return cnt;
}

int main(){
    int n;
    cout << "No of elements, n = ";
    cin >> n;
    int nums[n];
    cout << "Elements of nums = ";
    for (int i = 0; i < n; i ++){
        cin >> nums[i];
    }
    int target;
    cout << "target = ";
    cin >> target;
    cout << countPairs(nums, n, target) << endl;
    return 0;
}

File: Arrays/Q3. Count Pairs Whose Sum is Less than Target.py
class Solution:
    def countPairs(self, nums: list[int], target: int) -> int:
        res = 0
        nums.sort() 
        n = len(nums)
        
        left, right = 0, n - 1
        while left < right:
            if nums[left] + nums[right] < target:
                # If the sum is less than target, all pairs from left to right are valid
                res += (right - left) 
                left += 1  # Move the left pointer to the right
            else:
                right -= 1  # Move the right pointer to the left to try a smaller sum

        return res

sol = Solution()

nums = [-1, 1, 2, 3, 1]
target = 2
print(sol.countPairs(nums, target))  

nums = [-6, 2, 5, -2, -7, -1, 3]
target = -2
print(sol.countPairs(nums, target))  
